modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php

```diff
@@ -0,0 +1,247 @@
+	}
+
+	/**
+	 * Handles incoming request to run the batch of job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @throws Exception Invalid request for background process.
+	 */
+	public function handle_request() {
+		try {
+			// Exit if nonce varification fails.
+			$nonce_check = check_ajax_referer( self::BG_PROCESS_ACTION, 'nonce', false );
+
+			// If nonce check fails, fallback to key checking.
+			if ( false === $nonce_check ) {
+				throw new Exception( __( 'Invalid nonce passed to process.', 'performance-lab' ) );
+			}
+
+			$job_id = isset( $_REQUEST['job_id'] ) ? absint( sanitize_text_field( $_REQUEST['job_id'] ) ) : 0;
+
+			// Job ID is mandatory to be specified.
+			if ( empty( $job_id ) ) {
+				throw new Exception( __( 'No job specified to execute.', 'performance-lab' ) );
+			}
+
+			$this->job = perflab_get_background_job( $job_id );
+
+			// Silently exit if the job is not ready to run.
+			if ( ! $this->job->should_run() ) {
+				return;
+			}
+
+			$this->run(); // Run the job.
+
+		} catch ( Exception $e ) {
+			// @todo Add the error handling in separate issue.
+		} finally {
+			if ( $this->job instanceof Perflab_Background_Job ) {
+				// Unlock the process once everything is done.
+				$this->job->unlock();
+				$this->next_batch( $this->job->get_id() );
+			}
+		}
+	}
+
+	/**
+	 * Runs the process over a batch of job.
+	 *
+	 * This will trigger a custom action `perflab_job_{$job_action}` by hooking to which
+	 * consumer code can perform the relevant tasks. It will continue till:
+	 * 1. Allowed memory has been exceeded (90% of memory allotted to php script).
+	 * 2. Maximum execution time has been reached or exceeded. (10 seconds less than allotted time).
+	 * 3. Job is completed. Marking job as complete is responsibility of consumer code.
+	 *
+	 * @since n.e.x.t
+	 */
+	private function run() {
+		// Lock the process for this job before running.
+		$this->job->lock();
+
+		do {
+			/**
+			 * Consumer code will hook to this action to perform necessary tasks.
+			 *
+			 * @since n.e.x.t
+			 *
+			 * @param array $data Job data.
+			 */
+			do_action( 'perflab_job_' . $this->job->get_action(), $this->job->get_data() );
+		} while ( ! $this->memory_exceeded() && ! $this->time_exceeded( $this->job->get_id() ) && ! $this->job->is_completed() );
+	}
+
+	/**
+	 * Calls the next batch of the job if it is not completed already.
+	 *
+	 * This will send a POST request to admin-ajax.php with background
+	 * process specific action to continue executing the job in a new process.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param int $job_id Job ID. This is the term id from `background_job` taxonomy.
+	 */
+	private function next_batch( $job_id ) {
+		// Do not call the background process from within the script if the real cron has been setup to do so.
+		if ( defined( 'ENABLE_BG_PROCESS_CRON' ) ) {
+			return;
+		}
+
+		perflab_dispatch_background_process_request( $job_id );
```

Comment:
There's only a difference of `$key` in `start_job` and `next_batch` functions. I think we can somehow combine/compose these functions to avoid rewriting entire thing twice.